// sweet.js/dist/hygiene-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CollectBindingSyntax = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.collectBindings = collectBindings;

var _symbol = require("./symbol");

var _transforms = require("./transforms");

var _ramda = require("ramda");

var _ = _interopRequireWildcard(_ramda);

var _errors = require("./errors");

var _terms = require("./terms");

var _terms2 = _interopRequireDefault(_terms);

var _immutable = require("immutable");

var _astDispatcher = require("./ast-dispatcher");

var _astDispatcher2 = _interopRequireDefault(_astDispatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var CollectBindingSyntax = exports.CollectBindingSyntax = function (_ASTDispatcher) {
  _inherits(CollectBindingSyntax, _ASTDispatcher);

  function CollectBindingSyntax() {
    _classCallCheck(this, CollectBindingSyntax);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(CollectBindingSyntax).call(this, "collect", true));

    _this.names = (0, _immutable.List)();
    return _this;
  }

  _createClass(CollectBindingSyntax, [{
    key: "collect",
    value: function collect(term_322) {
      return this.dispatch(term_322);
    }
  }, {
    key: "collectBindingIdentifier",
    value: function collectBindingIdentifier(term_323) {
      return this.names.concat(term_323.name);
    }
  }, {
    key: "collectBindingPropertyIdentifier",
    value: function collectBindingPropertyIdentifier(term_324) {
      return this.collect(term_324.binding);
    }
  }, {
    key: "collectBindingPropertyProperty",
    value: function collectBindingPropertyProperty(term_325) {
      return this.collect(term_325.binding);
    }
  }, {
    key: "collectArrayBinding",
    value: function collectArrayBinding(term_326) {
      var _this2 = this;

      var restElement_327 = null;
      if (term_326.restElement != null) {
        restElement_327 = this.collect(term_326.restElement);
      }
      return this.names.concat(restElement_327).concat(term_326.elements.filter(function (el_328) {
        return el_328 != null;
      }).flatMap(function (el_329) {
        return _this2.collect(el_329);
      }));
    }
  }, {
    key: "collectObjectBinding",
    value: function collectObjectBinding(term_330) {
      var _this3 = this;

      return term_330.properties.flatMap(function (prop_331) {
        return _this3.collect(prop_331);
      });
    }
  }]);

  return CollectBindingSyntax;
}(_astDispatcher2.default);

function collectBindings(term_332) {
  return new CollectBindingSyntax().collect(term_332);
}
