// sweet.js/_site/build/test/assertions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testEval = exports.testParse = exports.stmt = exports.expr = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

exports.testParseFailure = testParseFailure;
exports.testModule = testModule;
exports.testThrow = testThrow;

var _sweet = require("../src/sweet");

var _expect = require("expect.js");

var _expect2 = _interopRequireDefault(_expect);

var _ramda = require("ramda");

var _babelCore = require("babel-core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function expr(program) {
  return stmt(program).expression;
}

function stmt(program) {
  return program.items[0];
}

function testParseFailure() {
  // TODO
}

function testParseWithOpts(code, acc, expectedAst, options) {
  var parsedAst = (0, _sweet.parse)(code, options);
  var isString = function isString(x) {
    return (0, _ramda.type)(x) === 'String';
  };
  var isObject = function isObject(x) {
    return (0, _ramda.type)(x) === 'Object';
  };
  var isArray = function isArray(x) {
    return (0, _ramda.type)(x) === 'Array';
  };

  function checkObjects(expected, actual) {
    var checkWithHygiene = (0, _ramda.cond)([[(0, _ramda.and)(isString, (0, _ramda.equals)('<<hygiene>>')), (0, _ramda.curry)(function (a, b) {
      return true;
    })], [isObject, (0, _ramda.curry)(function (a, b) {
      return checkObjects(a, b);
    })], [isArray, (0, _ramda.curry)(function (a, b) {
      return (0, _ramda.map)(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 2);

        var a = _ref2[0];
        var b = _ref2[1];
        return checkObjects(a, b);
      }, (0, _ramda.zip)(a, b));
    })], [_ramda.T, (0, _ramda.curry)(function (a, b) {
      return (0, _expect2.default)(b).to.be(a);
    })]]);

    (0, _ramda.mapObjIndexed)(function (prop, key, ob) {
      var checker = checkWithHygiene(prop);
      (0, _expect2.default)(actual).to.have.property(key);
      checker(actual[key]);
    }, expected);
  }
  checkObjects(expectedAst, acc(parsedAst));
}

function testModule(code, loader, expectedAst) {
  return testParseWithOpts(code, function (x) {
    return x;
  }, expectedAst, {
    loc: false,
    moduleResolver: function moduleResolver(x) {
      return x;
    },
    moduleLoader: function moduleLoader(path) {
      return loader[path];
    }
  });
}

// if a property has the string <<hygiene> it is ignored
function testParse(code, acc, expectedAst) {
  return testParseWithOpts(code, acc, expectedAst, {
    loc: false,
    moduleResolver: function moduleResolver() {
      return "";
    },
    moduleLoader: function moduleLoader() {
      return "";
    }
  });
}

function testEval(source, expectedOutput) {
  var result = (0, _sweet.compile)(source, { cwd: '.', transform: _babelCore.transform });
  var output;
  eval(result.code);
  (0, _expect2.default)(output).to.be(expectedOutput);
}

function testThrow(source) {
  (0, _expect2.default)(function () {
    return (0, _sweet.compile)(source, { cwd: '.', transform: _babelCore.transform });
  }).to.throwError();
}

exports.expr = expr;
exports.stmt = stmt;
exports.testParse = testParse;
exports.testEval = testEval;
